src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
+                Ok(Assigned::Done(Assignment {
+                    assigned: entry,
+                    assigned_to: buf,
+                    replaced,
+                }))
+            } else {
+                // if this is not the last token, we continue with a mutable
+                // reference to the entry as the next value
+                Ok(Assigned::Continue {
+                    same_src: src,
+                    next_buf: buf,
+                    next_value: entry,
+                })
+            }
+        }
+        Entry::Vacant(entry) => {
+            // if the entry does not exist, we create a value based on the
+            // remaining path with the src value as a leaf and assign it to the
+            // entry
+            let src = expand_src_path(remaining, src)?;
+            let assigned = entry.insert(src);
+            Ok(Assigned::Done(Assignment {
+                assigned,
+                assigned_to: buf,
+                replaced: None,
+            }))
+        }
+    }
+}
+
+fn assign_scalar<'v>(
+    token: Token<'_>,
+    remaining: &'_ Pointer,
+    mut buf: PointerBuf,
+    value: &'v mut Value,
+    src: Value,
+) -> Result<Assigned<'v>, AssignError> {
+    // scalar values are always replaced at the current buf (with its token)
+    // build the new src and we replace the value with it.
+
+    buf.push_back(token);
+    let src = expand_src_path(remaining, src)?;
+    let replaced = Some(mem::replace(value, src));
+
+    Ok(Assigned::Done(Assignment {
+        assigned: value,
+        assigned_to: buf,
+        replaced,
+    }))
+}
+
+fn expand_src_path(mut path: &Pointer, mut src: Value) -> Result<Value, AssignError> {
+    while let Some((ptr, tok)) = path.split_back() {
+        path = ptr;
+        match tok.decoded().as_ref() {
+            "0" | "-" => {
+                src = Value::Array(vec![src]);
+            }
+            _ => {
+                let mut obj = Map::new();
+                obj.insert(tok.to_string(), src);
+                src = Value::Object(obj);
+            }
+        }
+    }
+    Ok(src)
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::{json, Value};
+
+    use crate::{Pointer, PointerBuf};
+
+    #[test]
+    fn test_assign() {
+        let mut data = json!({});
+        let ptr = Pointer::from_static("/foo");
+        let val = json!("bar");
+
+        let assignment = ptr.assign(&mut data, val.clone()).unwrap();
+        assert_eq!(assignment.replaced, None);
+        assert_eq!(assignment.assigned, &val);
+        assert_eq!(assignment.assigned_to, "/foo");
+
+        // now testing replacement
+        let val2 = json!("baz");
+        let assignment = ptr.assign(&mut data, val2.clone()).unwrap();
+        assert_eq!(assignment.replaced, Some(Value::String("bar".to_string())));
+        assert_eq!(assignment.assigned, &val2);
+        assert_eq!(assignment.assigned_to, "/foo");
+    }
+
+    #[test]
+    fn test_assign_with_explicit_array_path() {
```

Comment:
Worth also checking what happens if:
1. The array exists, but is empty
2. The array has one element
3. The array has two elements
4. The path ends at the index

Could potentially cover this with a quickcheck test though.